docs/source/usecases/FunctionPointers.md

```diff
@@ -0,0 +1,286 @@
+
+If we do this `A` will contain a function pointer to a device function. That means we can capture `A` into a parallel region and execute it on the device, but now it will crash on the host.
+
+```c++
+#include <Kokkos_Core.hpp>
+#include <cmath>
+
+KOKKOS_INLINE_FUNCTION void foo() {
+  KOKKOS_IF_ON_HOST(printf("foo called from host\n");)
+  KOKKOS_IF_ON_DEVICE(printf("foo called from device\n");)
+}
+
+struct SomeClass {
+  void (*bar)();
+  KOKKOS_FUNCTION void print() const {
+    bar();
+  }
+};
+
+int main(int argc, char* argv[]) {
+  Kokkos::initialize(argc, argv);
+  {
+    SomeClass A;
+    Kokkos::View<SomeClass> A_v("A_v");
+    // Now init the function pointer on device:
+    Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>(0,1), KOKKOS_LAMBDA(int) {
+      A_v().bar = &foo;
+    });
+    // copy the device instance back to the host
+    Kokkos::deep_copy(A, A_v);
+
+    printf("Now I can capture A in a device kernel and use it there!\n");
+    Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>(0,1), KOKKOS_LAMBDA(int) {
+      A.print();
+    });
+    Kokkos::fence();
+    printf("But now I will crash on the host :-(\n");
+    A.print();
+    Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::DefaultHostExecutionSpace>(0,1), KOKKOS_LAMBDA(int) {
+      A.print();
+    });
+    Kokkos::fence();
+    printf("Never got here in CUDA/HIP builds\n");
+  }
+  Kokkos::finalize();
+}
+```
+
+If you run this with a CUDA build you get this output:
+
+```
+Now I can capture A in a device kernel and use it there!
+foo called from device
+But now I will crash on the host :-(
+Segmentation fault (core dumped)
+```
+
+## Creating a dual function pointer object
+
+To do better than this we need something which contains both the device function pointer and the host function pointer.
+And it should call the right thing depending on where you call it from.
+We can in-fact write such a class:
+
+```c++
+template<class FPtr>
+struct DualFunctionPtr {
+  FPtr h;
+  FPtr d;
+  template<class ... Args>
+  KOKKOS_FUNCTION
+  auto operator()(Args...args) const {
+    KOKKOS_IF_ON_HOST( return h(args...); )
+    KOKKOS_IF_ON_DEVICE( return d(args...); )
+  }
+};
+```
+
+This class is templated on the function pointer type, contains a pointer for both the host and the device version, and has a templated operator that forwards all the arguments, and calls the appropriate function pointer depending on call site.
+
+We can use this class similar to `std::function` inside `SomeClass`:
+
+```c++
+struct SomeClass {
+  DualFunctionPtr<decltype(&foo)> bar;
+  KOKKOS_FUNCTION void print() const {
+    bar();
+  }
+};
+```
+
+However we still need to initialize the two function pointers, using the device initialization approach from before, and also initializing the host side. Note: the order of initialization and `deep_copy` matters, because `deep_copy` will overwrite both member poitners in our wrapper class:
```

Comment:
```suggestion
However, we still need to initialize the two function pointers, using the device initialization approach from before, and also initializing the host side. Note: the order of initialization and `deep_copy` matters, because `deep_copy` will overwrite both member poitners in our wrapper class:
```